pkg/lib/synthesize/synthesis.go

```diff
@@ -0,0 +1,696 @@
+		structuredObjectMap:         map[string]*[]unstructured.Unstructured{},
+		logger:                      log.WithFields(log.Fields{"type": "synthesize"}),
+		apiServiceDescriptionMap:    map[v1.GroupVersionKind]*v1alpha1.APIServiceDescription{},
+		customResourceDefinitionMap: map[v1.GroupVersionKind]*v1alpha1.CRDDescription{},
+	}, nil
+}
+
+func (s *synthesize) GenerateCSV() (*v1alpha1.ClusterServiceVersion, error) {
+	if err := s.loadStructuredObjects(); err != nil {
+		return nil, err
+	}
+
+	if err := s.generateCsv(); err != nil {
+		if err == NoCsvInferanceFileFoundError && s.existingCSVPath == "" {
+			return nil, fmt.Errorf("a bundle should contain either a CSV or non-inferable CSV information in the %s directory", s.metadataDir)
+		}
+		return nil, err
+	}
+
+	return s.synthesizedCSV, nil
+}
+
+func (s *synthesize) generateCsv() error {
+	csvBasics, err := s.getNonInferableCSV()
+	if err != nil {
+		return err
+	}
+
+	installStrategy, err := s.getInstallStrategy()
+	if err != nil {
+		return err
+	}
+
+	specCRDs, err := s.getSpecCRDs()
+	if err != nil {
+		return err
+	}
+
+	apiServiceDefs, err := s.getAPIServiceDefinitions()
+	if err != nil {
+		return err
+	}
+
+	s.synthesizedCSV = &v1alpha1.ClusterServiceVersion{
+		TypeMeta: v1.TypeMeta{
+			APIVersion: v1alpha1.ClusterServiceVersionAPIVersion,
+			Kind:       CSVKind,
+		},
+		ObjectMeta: v1.ObjectMeta{
+			Name: csvBasics.Name,
+		},
+		Spec: v1alpha1.ClusterServiceVersionSpec{
+			InstallStrategy:           *installStrategy,
+			CustomResourceDefinitions: *specCRDs,
+			APIServiceDefinitions:     *apiServiceDefs,
+			Version:                   csvBasics.Version,
+			MinKubeVersion:            csvBasics.MinKubeVersion,
+			Maturity:                  csvBasics.Maturity,
+			DisplayName:               csvBasics.DisplayName,
+			Description:               csvBasics.Description,
+			Keywords:                  csvBasics.Keywords,
+			Maintainers:               csvBasics.Maintainers,
+			Provider:                  csvBasics.Provider,
+			Links:                     csvBasics.Links,
+			Icon:                      csvBasics.Icon,
+			InstallModes:              csvBasics.InstallModes,
+			Replaces:                  csvBasics.Replaces,
+			Labels:                    csvBasics.Labels,
+			Annotations:               csvBasics.Annotations,
+			Selector:                  csvBasics.Selector,
+		},
+	}
+	return nil
+}
+
+// loadStructuredObjects gets supported structured object recursively from bundle directory and put them into the map.
+// Map of the same bundle should always produce the same map considering directory walk is lexical ordered and therefore
+// input order is deterministic.
+// All supported type of structures are presented in supportedResources, extend the support types by adding to the map.
+func (s *synthesize) loadStructuredObjects() error {
+	return filepath.Walk(s.bundleDir, func(path string, info os.FileInfo, err error) error {
```

Comment:
So now that I've had more time to look at this, I think things would be less complicated if we forego the recursive search. Bundles are flat and we've already been given the `metadata` and `manifests` directories as arguments.